File: scripts/knowledge.ts
#!/usr/bin/env -S node --import tsx
/**
 * CLI for the trusted knowledge tree.
 *
 *   knowledge.ts check
 *   knowledge.ts resolve --query <text>
 *
 * `check` validates `knowledge/` against `literature/ref.bib` and exits 1 with
 * every diagnostic when anything is wrong. `resolve` prints one JSON document —
 * the reading bundle an agent must read, or the alternatives it must choose
 * between — and exits 0 for `match`, `ambiguous`, and `no-match` alike: those
 * are answers, not failures. Only an invocation the CLI cannot honour and an
 * invalid tree exit 1.
 *
 * There are deliberately no path or output overrides. The trusted tree of this
 * repository is the only thing this command may read, so an agent cannot be
 * talked into resolving against a directory someone else prepared.
 */

import path from "node:path";
import { fileURLToPath } from "node:url";
import { parseArgs } from "node:util";

import {
  KnowledgeValidationError,
  resolveKnowledge,
  validateKnowledge,
} from "../lib/knowledge/index.js";

const USAGE = [
  "usage:",
  "  knowledge.ts check",
  "  knowledge.ts resolve --query <text>",
].join("\n");

const REPO_ROOT = path.resolve(fileURLToPath(import.meta.url), "..", "..");

class UsageError extends Error {}

/** The `--query` of a `resolve` invocation, or a usage error explaining why not. */
function queryOf(args: readonly string[]): string {
  let values;
  try {
    ({ values } = parseArgs({
      args: [...args],
      options: { query: { type: "string" } },
      strict: true,
      allowPositionals: false,
    }));
  } catch (error) {
    throw new UsageError(error instanceof Error ? error.message : String(error));
  }
  if (values.query === undefined) {
    throw new UsageError("resolve requires --query <text>");
  }
  return values.query;
}

async function main(argv: readonly string[]): Promise<void> {
  const [command, ...rest] = argv;

  switch (command) {
    case "check": {
      if (rest.length > 0) {
        throw new UsageError("check takes no arguments");
      }
      const report = await validateKnowledge({ repoRoot: REPO_ROOT });
      if (!report.ok) {
        // The error's message is the whole report, so `check` and `resolve`
        // describe an invalid tree in exactly the same words.
        throw new KnowledgeValidationError(report);
      }
      console.log("knowledge: the trusted tree is valid");
      return;
    }
    case "resolve": {
      const result = await resolveKnowledge(queryOf(rest), { repoRoot: REPO_ROOT });
      console.log(JSON.stringify(result, null, 2));
      return;
    }
    case undefined:
      throw new UsageError("a subcommand is required");
    default:
      throw new UsageError(`unknown subcommand "${command}"`);
  }
}

try {
  await main(process.argv.slice(2));
} catch (error) {
  const message = error instanceof Error ? error.message : String(error);
  console.error(
    error instanceof UsageError ? `knowledge: ${message}\n${USAGE}` : `knowledge: ${message}`,
  );
  process.exitCode = 1;
}

File: scripts/knowledge.py
#!/usr/bin/env python3
"""Human- and agent-facing commands for trusted knowledge."""

from __future__ import annotations

import argparse
import json
from pathlib import Path
import subprocess
import sys

from lib.knowledge import (
    build_knowledge_site,
    preview_knowledge_site,
    resolve_knowledge,
    validate_knowledge,
)


ROOT = Path(__file__).resolve().parents[1]


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(prog="knowledge")
    subparsers = parser.add_subparsers(dest="command", required=True)
    check = subparsers.add_parser("check", help="validate trusted knowledge")
    check.add_argument(
        "--repo-root",
        type=Path,
        default=ROOT,
        help=argparse.SUPPRESS,
    )
    resolve = subparsers.add_parser("resolve", help="resolve a trusted reading bundle")
    resolve.add_argument("--query", required=True)
    resolve.add_argument(
        "--repo-root",
        type=Path,
        default=ROOT,
        help=argparse.SUPPRESS,
    )
    build = subparsers.add_parser(
        "build",
        help="safely render and atomically publish trusted knowledge",
    )
    build.add_argument(
        "--repo-root",
        type=Path,
        default=ROOT,
        help=argparse.SUPPRESS,
    )
    preview = subparsers.add_parser(
        "preview",
        help="preview the same validated execution-disabled projection",
    )
    preview.add_argument(
        "--repo-root",
        type=Path,
        default=ROOT,
        help=argparse.SUPPRESS,
    )
    return parser


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    if args.command == "check":
        report = validate_knowledge(args.repo_root)
        if not report.ok:
            for diagnostic in report.diagnostics:
                print(diagnostic, file=sys.stderr)
            return 1
        print(
            f"Knowledge valid: {len(report.graph.pages)} pages "
            f"across {len(report.graph.topics)} topics."
        )
        return 0
    if args.command == "resolve":
        try:
            result = resolve_knowledge(args.query, args.repo_root)
        except ValueError as error:
            print(error, file=sys.stderr)
            return 1
        print(json.dumps(result, ensure_ascii=False, indent=2))
        return 0
    if args.command == "build":
        try:
            output = build_knowledge_site(repo_root=args.repo_root)
        except (
            OSError,
            RuntimeError,
            ValueError,
            subprocess.SubprocessError,
        ) as error:
            print(error, file=sys.stderr)
            return 1
        relative_output = output.relative_to(args.repo_root)
        print(f"Knowledge site built: {relative_output.as_posix()}")
        return 0
    if args.command == "preview":
        try:
            preview_knowledge_site(repo_root=args.repo_root)
        except (
            OSError,
            RuntimeError,
            ValueError,
            subprocess.SubprocessError,
        ) as error:
            print(error, file=sys.stderr)
            return 1
        print("Knowledge preview stopped.")
        return 0
    raise AssertionError(f"unhandled command: {args.command}")


if __name__ == "__main__":
    raise SystemExit(main())
